### PR TITLE
Feature/implement orders (basic)

### PR DESCRIPTION
**Pull request by Somiel**
- [x] DB scheme development
- [x] Order management feature implementation
- [x] View templates development
- [x] Role dependent custom views feature implementation

### CODE REVIEW

File: app/assets/javascripts/orders.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (murderface13) [inline app/assets/javascripts/orders.coffee]: why do you need that blank files? Ask scaffold to get answer :smirk:

File: app/assets/stylesheets/orders.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the orders controller here.
```
Comment (murderface13) [inline app/assets/stylesheets/orders.scss]: remove blank files if you don't use its, pls

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -0,0 +1,97 @@
+# Orders management controller
+class OrdersController < ApplicationController
+  before_action :set_order, only: [
+    :edit,
+    :update,
+    :destroy
+  ]
+
+  before_action :set_orders, only: :index
+  before_action :authenticate_user!, only: [
```
Comment (murderface13) [inline app/controllers/orders_controller.rb]: why you don't authenticate user for create action for example?

we should authentificate user in every controller actions and skip it if needs, i suppose

File: app/controllers/orders_controller.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    :new,
+    :index,
+    :edit,
+    :update,
+    :destroy
+  ]
+
+  # GET /admin/orders
+  def index
```
Comment (murderface13) [inline app/controllers/orders_controller.rb]: you don't need specify blank controller actions for CRUD(as it did scaffold). CRUD actions would be inherited from Action controller through Application controller

File: app/controllers/orders_controller.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  def set_order
+    @order = Order.find(params[:id])
+  end
+
+  def order_params
+    params.require(:order).permit(:starting_point, :destination_point, :status, :comment, :client_id)
+  end
+
+  def set_orders
+    if current_user.is_a? Manager
```
Comment (murderface13) [inline app/controllers/orders_controller.rb]: what about case-when operator?

File: app/models/order.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
+# This class represents Order model
+class Order < ActiveRecord::Base
+  belongs_to :client, foreign_key: 'client_id'
+  belongs_to :driver, foreign_key: 'driver_id'
+
+  STATUS_ACTIVE = 1
```
Comment (murderface13) [inline app/models/order.rb]: :scream: 
please, look at http://edgeapi.rubyonrails.org/classes/ActiveRecord/Enum.html


also, if you would have specific logic with order status changing - you might use smth like that: https://github.com/pluginaweek/state_machine

File: app/models/order.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  validates :status,
+            presence: true,
+            inclusion: [
+              Order::STATUS_ACTIVE,
+              Order::STATUS_PENDING,
+              Order::STATUS_COMPLETED,
+              Order::STATUS_CANCELED
+            ]
+
+  def status_name
```
Comment (murderface13) [inline app/models/order.rb]: brr :alien:

File: app/views/home/index.html.slim [hunk 8]
```diff
@@ -1,2 +1,18 @@
...
+  | elit placerat, vel varius massa cursus. Donec id ex lorem. Donec porttitor lacus eget velit consectetur
+  | suscipit. Fusce non fringilla velit. Cras congue scelerisque ultricies. Duis commodo urna eu neque
+  | imperdiet, eget placerat magna venenatis.
+
+
+h4.center-align class="greetings" Create an order
+
+.row
+  .col.s6.center-align style="margin: 0 25%;"
+    = render '/orders/form'
```
Comment (murderface13) [inline app/views/home/index.html.slim]: :smiling_imp:

File: app/views/orders/new.html.slim [hunk 9]
```diff
@@ -0,0 +1,4 @@
+h2.center-align Create order
+.row
+  .col.s6.center-align style="margin: 0 25%;"
+    = render "form"
```
Comment (murderface13) [inline app/views/orders/new.html.slim]: there aren't any new lines in the end of file

File: app/views/orders/_form.html.slim [hunk 10]
```diff
@@ -0,0 +1,28 @@
+= form_for(@order, html: { method: action_name == "edit"? :put : :post }) do |f|
+
+  - if flash[:error]
```
Comment (murderface13) [inline app/views/orders/_form.html.slim]: you render flashes in https://github.com/Somiel/SlobodaTaxi/blob/master/app/views/shared/_flash.html.slim , aren't you?

File: app/views/home/index.html.slim (hunk 8 above)
Comment (Somiel) [inline app/views/home/index.html.slim]: That was just a dummy)

File: app/assets/javascripts/orders.coffee (hunk 1 above)
Comment (Somiel) [inline app/assets/javascripts/orders.coffee]: just wanted to reserve ot for WS js code

File: app/controllers/orders_controller.rb (hunk 3 above)
Comment (Somiel) [inline app/controllers/orders_controller.rb]: i didnt require auth for create action to preserve order create feature for unregistered users,

y, i can skip before action callback

File: app/controllers/orders_controller.rb (hunk 4 above)
Comment (Somiel) [inline app/controllers/orders_controller.rb]: got it

File: app/views/orders/_form.html.slim (hunk 10 above)
Comment (Somiel) [inline app/views/orders/_form.html.slim]: Right, fixed

File: app/controllers/orders_controller.rb (hunk 5 above)
Comment (Somiel) [inline app/controllers/orders_controller.rb]: refactored

File: app/views/orders/new.html.slim (hunk 9 above)
Comment (Somiel) [inline app/views/orders/new.html.slim]: seems to be fixed

File: app/models/order.rb (hunk 7 above)
Comment (Somiel) [inline app/models/order.rb]: refactored

File: app/models/order.rb (hunk 6 above)
Comment (Somiel) [inline app/models/order.rb]: wow..haven`t ever seen it. nice stuff

about state machine idk yet..gonna do some research

File: app/models/driver.rb [hunk 11]
```diff
@@ -1,3 +1,8 @@
 # Driver user model
 class Driver < User
+  has_many :orders, foreign_key: 'driver_id'
+
+  def name_with_initial
```
Comment (murderface13) [inline app/models/driver.rb]: that method usually named 'full_name'

File: config/initializers/state_machine.rb [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+  module Integrations
+    module ActiveModel
+      public :around_validation
+    end
+
+    module ActiveRecord
+      public :around_save
+    end
+  end
+end
```
Comment (murderface13) [inline config/initializers/state_machine.rb]: what about new line? :smile:

Comment (Somiel) [inline config/initializers/state_machine.rb]: damn it, fixed))

**Comment by murderface13** [comment]
good job! :+1:  but what about some tests?

**Comment by Somiel** [comment]
Gotta l2RSpec at 1st :smile:

**Pull request merged**